src/Http/Http/src/StreamPipeReader.cs

```diff
@@ -0,0 +1,396 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Runtime.ExceptionServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Http
+{
+    /// <summary>
+    /// Implements PipeReader using an underlying stream.
+    /// </summary>
+    public class StreamPipeReader : PipeReader
+    {
+        private readonly int _minimumSegmentSize;
+        private readonly Stream _readingStream;
+        private readonly MemoryPool<byte> _pool;
+
+        private CancellationTokenSource _internalTokenSource;
+        private bool _isCompleted;
+        private ExceptionDispatchInfo _exceptionInfo;
+        private object lockObject = new object();
+
+        private BufferSegment _readHead;
+        private int _readIndex;
+
+        private BufferSegment _commitHead;
+        private long _consumedLength;
+        private bool _examinedEverything;
+
+        private CancellationTokenSource InternalTokenSource
+        {
+            get
```

Comment:
Could be written as:

```suggestion
            get => _internalTokenSource
                ?? Interlocked.CompareExchange(ref _internalTokenSource, new CancellationTokenSource(), null)
                ?? _internalTokenSource;
```

and together with the other suggestions one could get rid of `lockObject`.

Reply:
I dig it. 👍